items.md

```diff
@@ -534,6 +534,62 @@ Return codes:
 * 204 No Content - if the specified item is not yet versioned
 * 400 Bad Request - if the item id param is missing or invalid (not an uuid)
 
+### Get and create item identifiers
+**GET /api/core/items/{:item-uuid}/identifiers**
+
+Returns information about the identifiers associated with this item, for example Handle and DOI URIs. If relevant, the status of the identifier is also included.
+
+The JSON response is formatted like the example below (the same data model as the [identifiers submission step](submissionsections.md)).
+```json
+{
+  "identifiers" : [ {
+    "value" : "https://doi.org/10.33515/dspace-61",
+    "identifierType" : "doi",
+    "identifierStatus" : "TO_BE_REGISTERED",
+    "type" : "identifier"
+  }, {
+    "value" : "123456789/418",
+    "identifierType" : "handle",
+    "identifierStatus" : null,
+    "type" : "identifier"
+  } ],
+  "type" : "identifiers",
+  "_links" : {
+    "self" : {
+      "href" : "http://localhost:8080/server/api/core/items/6bea2772-0e71-4636-8c1c-5c132821fa2c/identifiers"
+    }
+  }
+}
+```
+Return codes:
+* 200 OK - if the operation succeeds
+* 400 Bad Request - if the item id param is missing or invalid (not an uuid)
+* 401 Unauthorized - if you are not authenticated and versioning is not public
+* 403 Forbidden - if you are not logged in with sufficient permissions and versioning is not public
+* 404 Not found - if the item doesn't exist
+
+**POST /api/core/items/{:item-uuid}/identifiers**
+
+Creates or registers a new identifier for this item. A 'type' parameter is required to indicate which sort of
+identifier to create, eg. doi, handle, or other supported identifier types.
+This operation would typically be used to mint and queue a DOI for registration for an item that is already archived.
+
+No body data is required in the POST. Example request:
+
+```
+ curl -i -X POST https://dspace7.4science.it/dspace-spring-rest/api/core/items/1911e8a4-6939-490c-b58b-a5d70f8d91fb/identifiers?type=doi 
+```
+
+On success, the item resource is returned as a response.
+
+Return codes:
+* 201 Created - if the operation succeeds
+* 302 Found - if an identifier registration was requested but the identifier is already registered
+* 400 Bad Request - if the item id param is missing or invalid (not an uuid), or if the type param is missing or invalid
+* 401 Unauthorized - if you are not authenticated and versioning is not public
+* 403 Forbidden - if you are not logged in with sufficient permissions and versioning is not public
+* 404 Not found - if the item doesn't exist
+
```

Comment:
@kshepherd these have been added after my initial review and make the scope wider than what originally expected.
A `/identifiers` path in the `/core/items/:uuid` endpoint should be an association endpoint (as described in our README.md), i.e. an endpoint to manage relations among the item specified by the `:uuid` and other rest resources.
This mean that we should define a REST resource that represent the pid and here just have the linkage. Please note that we already have an endpoint to deal with PIDs, the `/api/pid/find` that I just noted to be not documented at all in our REST contract...
I would suggest to design this endpoint around `/api/pid/identifiers` so that we can create new identifier just using the usual POST to this collection endpoint where the item is sent as uri-list and the type of identifier is specified by a type parameter. This would make much easier to delete or update an identifier (where it makes sense) and the lookup for all the identifiers of an item would be a standard search method.
In future I can also see other search methods to retrieve all the identifiers in a specific state and/or for a specific provider, such as all the pending DOIs, a such endpoint would be required for an administrative dashboard.

Reply:
@abollini Thank you for your feedback. For better understanding: why should we send a list of items to an identifiers endpoint instead of sending a post request to the item endpoint to create a new identifier? This is how e.g. bundles are being created currently. What is the difference between identifiers and bundles? To be sure, to understand you correctly, are you asking us to move the POST operation we have in the contract above from `/api/core/items/{:item-uuid}/identifiers` to `api/pid/identifiers`?

Identifiers in the sense of persistent identifiers (handles and dois) should never be changed and better not be deleted. Therefore, we would say pids are being deleted when an item is deleted and do not need an endpoint for deletion on their own.

The difference between `api/pid/find` and `api/core/items/{:item-uuid}/identifiers` is that the first one resolves an identifiers to an item and the second one lists all identifiers of an item.

We did not change the scope of the PR, we just missed to document this REST endpoint, while it already was in the backend PR. Sorry for that.

@abollini with the argument of treating creation of identifiers comparable to bitstreams, would you agree that we can leave the Post request were it is? With the explanation about what is searched (the dspace object identified or all identifiers of an item), can we leave the Get action where we put it?

Reply:
@pnbecker : I think I see what @abollini is saying... and I didn't notice it initially.

He's pointing out that we usually create a new "object" by POST to its own "resource" endpoint.  So, for example, here's how you create some major objects:
* Community: `POST /api/core/communities` (with an optional `parent` param)
* Collection: `POST /api/core/collections?parent=[:communityUUID]`
* Item (when bypassing workflow): `POST /api/core/items?owningCollection=[:collectionUUID]`

So, he's pointing out that creating an Identifier looks odd here... because we are doing:
`POST /api/core/items/{:item-uuid}/identifiers`

This makes "Identifier" seem **specific to Items** (and it's obviously not).  It also means that, if we ever wanted to create Collection Identifiers. we'd have to create a _new endpoint_ that does the same thing for Collections:
`POST /api/core/collections/{:collection-uuid}/identifiers`

That starts to get very repetitive very quickly, and it not very RESTful design.

Instead, a more generic approach is to create an **identifier** endpoint which is generic for all objects.  We already do have an example at `/api/pid/find`... as that endpoint locates any `DSpaceObject` via an identifier.

So, I think what @abollini is suggesting is to consider changing from this:
```
POST /api/core/itemss/{:item-uuid}/identifiers?type=doi
```
Into something like this:
```
POST /api/pid/identifiers?object=[:dsoUUID]&type=doi
```

You'd still *keep* the `GET /api/core/items/{:item-uuid}/identifiers` endpoint as it is designed.  That endpoint would still return all identifiers "linked" to that specific Item.  But, you'd change the *creation* process to be more generic... it'd work for any DSpaceObject type (eventually, it need not work for any DSO right now, we could just start with Items).

Reply:
thank you @tdonohue that makes perfect sense. :-) We'll change it.

Reply:
@abollini I've started attempted this refactor.

As far as I can see, RestRepository classes that implement `createAndReturn` accepting a uri-list are all classes that can implement `DSpaceRestRepository<SomethingRest, Integer>` or so on -- in other words, they all look like addressable resources. We don't have this concept in identifiers - we have two tables. handle and DOI, each with their own integer IDs.
I have an `IdentifiersRest` and `IdentifierRest` resource for mapping but they don't have additional IDs and I'm not sure how they fit into an DSpaceObjectRestRepository impl.

Are there other ways to accept uri-list without having to respond with a generic IdentifierRest object with addressable Ids? Or, in this case, can I just somehow use a 'string' for the ID, and the handle or DOI value becomes this ID that can then be used with /api/pid/find?

Reply:
hi @kshepherd,
the ID of the identifier resource could be a string like <identifier-type>:<identifier-value> this would be similar to what is done for the Vocabulary Entry Details resource, see https://github.com/DSpace/RestContract/blob/main/vocabularyEntryDetails.md#single-authority

The identifier resource must be addressable as we want it to be exposed under a dedicated endpoint i.e. `/api/pid/identifiers/doi:10.123%2Ftest01` please note that the / in the doi have been url encoded
the uri-list would represent the items object so right now you should not need to implement a dedicated `UriListHandler` but you should be able to reuse `org.dspace.app.rest.repository.handler.ExternalSourceItemUriListHandler`.
Maybe the `org.dspace.app.rest.repository.ClaimedTaskRestRepository.createAndReturn(Context, List<String>)` gives you a good example where a Repository responsible for a rest resource (ClaimedTask) is getting an uri-list of a different type of resource (PoolTask).

The current IdentifierController is an exception to our architecture, for your need you should be just able to create RestRepository to deal with GET, POST, DELETE. 
You will need to have a `@SearchRestMethod` in the IdentifierRestRepository to implemente the lookup by the dso uuid, it could be named findByDSO with a uuid parameter.

An additional search method in future could allow us to get rid of the `/api/pid/identifiers/find` endpoint but this is imho out-of-the-scope for this PR

Reply:
thanks @abollini, I will work towards this

Reply:
Hi @abollini , I have pushed (incomplete, some docs and work needed) changes that use /api/pid/identifiers for create. I also implemented findOne for this.
I had some issues:
- If I return 'not implemented' for findAll, it seems to break some HAL things with sub-methods like /search?
- I think I *do* need to incorporate the undocumented /api/pid/find functionality as another search method because the request routing doesn't like my IdentifierRestRepository using the same path as that controller
- Everything else should be OK. Identifiers can still be read as embedded item links, but the actual creation is done with a uri-list of item URIs POSTed to /api/pid/identifiers?type=doi

This is pushed with working angular code (the angular code as a few issues to fix up)